cpp/src/io/parquet/writer_impl.cu

```diff
@@ -695,148 +900,49 @@ void writer::impl::write(table_view const &table)
-  std::vector<parquet_column_view> parquet_columns;
-  parquet_columns.reserve(num_columns);  // Avoids unnecessary re-allocation
-
-  // because the repetition type is global (in the sense of, not per-rowgroup or per write_chunk()
-  // call) we cannot know up front if the user is going to end up passing tables with nulls/no nulls
-  // in the multiple write_chunk() case.  so we'll do some special handling.
-  // The user can pass in information about the nullability of a column to be enforced across
-  // write_chunk() calls, in a flattened bool vector. Figure out that per column.
-  auto per_column_nullability =
-    (single_write_mode)
-      ? std::vector<std::vector<bool>>{}
-      : get_per_column_nullability(table, user_metadata_with_nullability.column_nullable);
-
-  uint decimal_precision_idx = 0;
-
-  for (auto it = table.begin(); it < table.end(); ++it) {
-    const auto col        = *it;
-    const auto current_id = parquet_columns.size();
-
-    // if the user is explicitly saying "I am only calling this once", assume the columns in this
-    // one table tell us everything we need to know about their nullability.
-    // Empty nullability means the writer figures out the nullability from the cudf columns.
-    auto const &this_column_nullability =
-      (single_write_mode) ? std::vector<bool>{} : per_column_nullability[current_id];
-
-    parquet_columns.emplace_back(current_id,
-                                 col,
-                                 this_column_nullability,
-                                 user_metadata,
-                                 int96_timestamps,
-                                 decimal_precision,
-                                 decimal_precision_idx,
-                                 stream);
+  auto tbl_meta = (table_meta == nullptr) ? table_input_metadata(table) : *table_meta;
+  // Fill unnamed columns' names in tbl_meta
+  std::function<void(column_in_metadata &, std::string)> add_default_name =
+    [&](column_in_metadata &col_meta, std::string default_name) {
+      if (col_meta.get_name().empty()) col_meta.set_name(default_name);
+      for (size_type i = 0; i < col_meta.num_children(); ++i) {
+        add_default_name(col_meta.child(i), col_meta.get_name() + "_" + std::to_string(i));
+      }
+    };
+  for (size_t i = 0; i < tbl_meta.column_metadata.size(); ++i) {
+    add_default_name(tbl_meta.column_metadata[i], "_col" + std::to_string(i));
   }
 
-  CUDF_EXPECTS(decimal_precision_idx == decimal_precision.size(),
-               "Too many decimal precision values!");
-
-  // first call. setup metadata. num_rows will get incremented as write_chunk is
-  // called multiple times.
-  // Calculate the sum of depths of all list columns
-  size_type const list_col_depths = std::accumulate(
-    parquet_columns.cbegin(), parquet_columns.cend(), 0, [](size_type sum, auto const &col) {
-      return sum + col.nesting_levels();
-    });
+  auto vec         = input_table_to_linked_columns(table);
```

Comment:
could we own the linked columns here and just keep pointers in the schema?

Reply:
That was my first attempt. https://github.com/rapidsai/cudf/pull/7461/commits/48e7ad4c9ea88997db1089e70c4ed8791454138d. 

The linked columns have to be shared pointers because for the child to have a pointer to its parent, the parent's pointer should not change. That means that `vec` has to be a vector of shared pointers. Once these are shared pointers anyway, it made sense to replace the raw pointer in the schema with a shared pointer.